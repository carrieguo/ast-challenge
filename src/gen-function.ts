import * as t from "@babel/types";

export function genFunction(
  responseType: string,
  paramTypeName: string,
  functionName: string,
  keyName = "poolsQuery",
  queryServiceMethodName = "pools"
) {
  // 创建函数参数
  const requestParam = t.objectPattern([
    t.objectProperty(
      t.identifier("request"),
      t.identifier("request"),
      false,
      true
    ),
    t.objectProperty(
      t.identifier("options"),
      t.identifier("options"),
      false,
      true
    ),
  ]);

  requestParam.typeAnnotation = t.tsTypeAnnotation(
    t.tsTypeReference(
      t.identifier(paramTypeName),
      t.tsTypeParameterInstantiation([t.tsTypeReference(t.identifier("TData"))])
    )
  );

  const functionBody = t.blockStatement([
    t.returnStatement(
      t.callExpression(t.identifier("useQuery"), [
        t.arrayExpression([t.stringLiteral(keyName), t.identifier("request")]),
        t.arrowFunctionExpression(
          [],
          t.blockStatement([
            t.ifStatement(
              t.unaryExpression("!", t.identifier("queryService")),
              t.blockStatement([
                t.throwStatement(
                  t.newExpression(t.identifier("Error"), [
                    t.stringLiteral("Query Service not initialized"),
                  ])
                ),
              ])
            ),
            t.returnStatement(
              t.callExpression(
                t.memberExpression(
                  t.identifier("queryService"),
                  t.identifier(queryServiceMethodName)
                ),
                [t.identifier("request")]
              )
            ),
          ])
        ),
      ])
    ),
  ]);
  const functionExpression = t.arrowFunctionExpression(
    [requestParam],
    functionBody,
    true
  );

  // 声明泛型类型参数 TData
  functionExpression.typeParameters = t.tsTypeParameterDeclaration([
    t.tsTypeParameter(
      null,
      t.tsTypeReference(t.identifier(responseType)),
      "TData"
    ),
  ]);

  const constDec = t.variableDeclaration("const", [
    t.variableDeclarator(t.identifier(functionName), functionExpression),
  ]);
  return constDec;
}
