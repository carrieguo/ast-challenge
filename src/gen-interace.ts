import * as t from "@babel/types";

export function genInterface(
  requestType: string,
  responseType: string,
  interaceName: string
) {
  const typeParameters = t.tsTypeParameterDeclaration([
    t.tsTypeParameter(null, undefined, "TData"),
  ]);

  const _extends = [
    t.tsExpressionWithTypeArguments(
      t.identifier("ReactQueryParams"),
      t.tsTypeParameterInstantiation([
        t.tsTypeReference(t.identifier(responseType)),
        t.tsTypeReference(t.identifier("TData")),
      ])
    ),
  ];

  const requestProp = t.tsPropertySignature(
    t.identifier("request"),
    t.tsTypeAnnotation(t.tsTypeReference(t.identifier(requestType)))
  );

  requestProp.optional = true;

  const interfaceBody = t.tsInterfaceBody([requestProp]);

  return t.tsInterfaceDeclaration(
    t.identifier(interaceName),
    typeParameters,
    _extends,
    interfaceBody
  );
}
