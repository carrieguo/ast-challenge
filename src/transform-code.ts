import babelTraverse from "@babel/traverse";
import { parse, ParserPlugin } from "@babel/parser";
import generate from "@babel/generator";
import * as t from "@babel/types";

const code = `
export interface INTERFACE_NAME<TData> extends ReactQueryParams<RESPONSE_TYPE, TData> {
    request?: REQUEST_TYPE;
}
const FUNCTION_NAME = <TData = RESPONSE_TYPE,>({
    request,
    options
}: INTERFACE_NAME<TData>) => {
    return useQuery<RESPONSE_TYPE, Error, TData>(["KEY_NAME", request], () => {
        if (!queryService) throw new Error("Query Service not initialized");
        return queryService.QUERY_SERVICE_METHOD_NAME(request);
    }, options);
};
`;

export function transformCode(
  requestType: string,
  responseType: string,
  paramTypeName: string,
  functionName: string,
  keyName = "poolsQuery",
  queryServiceMethodName = "pools"
) {
  const plugins: ParserPlugin[] = ["typescript"];

  const ast = parse(code, {
    sourceType: "module",
    plugins,
  });

  babelTraverse(ast as any, {
    TSInterfaceDeclaration(path) {
      path.node.id.name = paramTypeName;
    },
    Identifier(path) {
      if (path.node.name === "REQUEST_TYPE") {
        path.node.name = requestType;
      } else if (path.node.name === "RESPONSE_TYPE") {
        path.node.name = responseType;
      } else if (path.node.name === "FUNCTION_NAME") {
        path.node.name = functionName;
      } else if (path.node.name === "QUERY_SERVICE_METHOD_NAME") {
        path.node.name = queryServiceMethodName;
      }
    },
    StringLiteral(path) {
      if (path.node.value === "KEY_NAME") {
        path.node.value = keyName;
      }
    },
  });

  return ast;
}
