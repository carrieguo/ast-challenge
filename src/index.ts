import * as t from "@babel/types";
import { genInterface } from "../src/gen-interace";
import { genFunction } from "../src/gen-function";
import { transformCode } from "./transform-code";

export default (
  obj: Record<string, { requestType: string; responseType: string }>,
  type: 1 | 2 = 1
) => {
  return t.file(
    t.program([
      ...Object.entries(obj).flatMap(([key, value]) => {
        const functionName = `use${key}Query`;
        const interfaceName = `Use${key}Query`;
        const { requestType, responseType } = obj[key];
        if (type === 1) {
          // 使用pure，直接生成
          return [
            genInterface(requestType, responseType, interfaceName),
            genFunction(responseType, interfaceName, functionName),
          ];
        } else {
          // 使用transformCode
          return transformCode(
            requestType,
            responseType,
            interfaceName,
            functionName
          ).program.body;
        }
      }),
    ])
  );
};
