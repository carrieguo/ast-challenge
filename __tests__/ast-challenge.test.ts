import generate from "@babel/generator";
import * as t from "@babel/types";
import genCode from "../src";
import { transformCode } from "../src/transform-code";
import exampleMethods from "../example-methods.json";
const expectCode = (ast) => {
  expect(generate(ast).code).toMatchSnapshot();
};

it("gen-code", () => {
  expectCode(genCode(exampleMethods, 1));
});

it("transform-code", () => {
  expectCode(genCode(exampleMethods, 1));
});
