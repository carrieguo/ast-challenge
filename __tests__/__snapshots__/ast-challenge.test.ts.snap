// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gen-code 1`] = `
"interface UsePoolsQuery<TData> extends ReactQueryParams<QueryPoolsResponse, TData> {
  request?: QueryPoolsRequest;
}

const usePoolsQuery = async <TData = QueryPoolsResponse,>({
  request,
  options
}: UsePoolsQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseNumPoolsQuery<TData> extends ReactQueryParams<QueryNumPoolsResponse, TData> {
  request?: QueryNumPoolsRequest;
}

const useNumPoolsQuery = async <TData = QueryNumPoolsResponse,>({
  request,
  options
}: UseNumPoolsQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseTotalLiquidityQuery<TData> extends ReactQueryParams<QueryTotalLiquidityResponse, TData> {
  request?: QueryTotalLiquidityRequest;
}

const useTotalLiquidityQuery = async <TData = QueryTotalLiquidityResponse,>({
  request,
  options
}: UseTotalLiquidityQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UsePoolQuery<TData> extends ReactQueryParams<QueryPoolResponse, TData> {
  request?: QueryPoolRequest;
}

const usePoolQuery = async <TData = QueryPoolResponse,>({
  request,
  options
}: UsePoolQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UsePoolParamsQuery<TData> extends ReactQueryParams<QueryPoolParamsResponse, TData> {
  request?: QueryPoolParamsRequest;
}

const usePoolParamsQuery = async <TData = QueryPoolParamsResponse,>({
  request,
  options
}: UsePoolParamsQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseTotalPoolLiquidityQuery<TData> extends ReactQueryParams<QueryTotalPoolLiquidityResponse, TData> {
  request?: QueryTotalPoolLiquidityRequest;
}

const useTotalPoolLiquidityQuery = async <TData = QueryTotalPoolLiquidityResponse,>({
  request,
  options
}: UseTotalPoolLiquidityQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseTotalSharesQuery<TData> extends ReactQueryParams<QueryTotalSharesResponse, TData> {
  request?: QueryTotalSharesRequest;
}

const useTotalSharesQuery = async <TData = QueryTotalSharesResponse,>({
  request,
  options
}: UseTotalSharesQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseSpotPriceQuery<TData> extends ReactQueryParams<QuerySpotPriceResponse, TData> {
  request?: QuerySpotPriceRequest;
}

const useSpotPriceQuery = async <TData = QuerySpotPriceResponse,>({
  request,
  options
}: UseSpotPriceQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseEstimateSwapExactAmountInQuery<TData> extends ReactQueryParams<QuerySwapExactAmountInResponse, TData> {
  request?: QuerySwapExactAmountInRequest;
}

const useEstimateSwapExactAmountInQuery = async <TData = QuerySwapExactAmountInResponse,>({
  request,
  options
}: UseEstimateSwapExactAmountInQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseEstimateSwapExactAmountOutQuery<TData> extends ReactQueryParams<QuerySwapExactAmountOutResponse, TData> {
  request?: QuerySwapExactAmountOutRequest;
}

const useEstimateSwapExactAmountOutQuery = async <TData = QuerySwapExactAmountOutResponse,>({
  request,
  options
}: UseEstimateSwapExactAmountOutQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};"
`;

exports[`transform-code 1`] = `
"interface UsePoolsQuery<TData> extends ReactQueryParams<QueryPoolsResponse, TData> {
  request?: QueryPoolsRequest;
}

const usePoolsQuery = async <TData = QueryPoolsResponse,>({
  request,
  options
}: UsePoolsQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseNumPoolsQuery<TData> extends ReactQueryParams<QueryNumPoolsResponse, TData> {
  request?: QueryNumPoolsRequest;
}

const useNumPoolsQuery = async <TData = QueryNumPoolsResponse,>({
  request,
  options
}: UseNumPoolsQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseTotalLiquidityQuery<TData> extends ReactQueryParams<QueryTotalLiquidityResponse, TData> {
  request?: QueryTotalLiquidityRequest;
}

const useTotalLiquidityQuery = async <TData = QueryTotalLiquidityResponse,>({
  request,
  options
}: UseTotalLiquidityQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UsePoolQuery<TData> extends ReactQueryParams<QueryPoolResponse, TData> {
  request?: QueryPoolRequest;
}

const usePoolQuery = async <TData = QueryPoolResponse,>({
  request,
  options
}: UsePoolQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UsePoolParamsQuery<TData> extends ReactQueryParams<QueryPoolParamsResponse, TData> {
  request?: QueryPoolParamsRequest;
}

const usePoolParamsQuery = async <TData = QueryPoolParamsResponse,>({
  request,
  options
}: UsePoolParamsQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseTotalPoolLiquidityQuery<TData> extends ReactQueryParams<QueryTotalPoolLiquidityResponse, TData> {
  request?: QueryTotalPoolLiquidityRequest;
}

const useTotalPoolLiquidityQuery = async <TData = QueryTotalPoolLiquidityResponse,>({
  request,
  options
}: UseTotalPoolLiquidityQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseTotalSharesQuery<TData> extends ReactQueryParams<QueryTotalSharesResponse, TData> {
  request?: QueryTotalSharesRequest;
}

const useTotalSharesQuery = async <TData = QueryTotalSharesResponse,>({
  request,
  options
}: UseTotalSharesQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseSpotPriceQuery<TData> extends ReactQueryParams<QuerySpotPriceResponse, TData> {
  request?: QuerySpotPriceRequest;
}

const useSpotPriceQuery = async <TData = QuerySpotPriceResponse,>({
  request,
  options
}: UseSpotPriceQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseEstimateSwapExactAmountInQuery<TData> extends ReactQueryParams<QuerySwapExactAmountInResponse, TData> {
  request?: QuerySwapExactAmountInRequest;
}

const useEstimateSwapExactAmountInQuery = async <TData = QuerySwapExactAmountInResponse,>({
  request,
  options
}: UseEstimateSwapExactAmountInQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};

interface UseEstimateSwapExactAmountOutQuery<TData> extends ReactQueryParams<QuerySwapExactAmountOutResponse, TData> {
  request?: QuerySwapExactAmountOutRequest;
}

const useEstimateSwapExactAmountOutQuery = async <TData = QuerySwapExactAmountOutResponse,>({
  request,
  options
}: UseEstimateSwapExactAmountOutQuery<TData>) => {
  return useQuery([\\"poolsQuery\\", request], () => {
    if (!queryService) {
      throw new Error(\\"Query Service not initialized\\");
    }

    return queryService.pools(request);
  });
};"
`;
